### ISSUE TITLE
Unable to connect to Firestore emulator using testing app instance

### ISSUE THREAD

**Issue by the-jackalope** [firebase/firebase-js-sdk#2619]
### Describe your environment

  * Operating System version: 
  * Browser version: macOS Cataline 10.15.2
  * Firebase SDK version:
```
"firebase": "^6.4.1",
"firebase-admin": "^8.9.2",
"firebase-functions": "^3.3.0",
...
"@firebase/testing": "^0.16.8",
"firebase-functions-test": "^0.1.7",
```
  * Firebase Product: database

### Describe the problem
I'm trying to run tests locally using the Firestore emulator, but am completely unable to connect. I consistently get the error message:
```
  console.error node_modules/@firebase/testing/node_modules/@firebase/logger/dist/index.cjs.js:94
    [2020-02-11T22:24:18.371Z]  @firebase/firestore: Firestore (7.8.1): Could not reach Cloud Firestore backend. Connection failed 1 times. Most recent error: FirebaseError: [code=unauthenticated]: 16 UNAUTHENTICATED: Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.
    This typically indicates that your device does not have a healthy Internet connection at the moment. The client will operate in offline mode until it is able to successfully connect to the backend.


Error: 14 UNAVAILABLE: failed to connect to all addresses

    at Object.exports.createStatusError (/Users/johngcook/Developer/dex/dex-core/cf_node/functions/node_modules/grpc/src/common.js:91:15)
    at Object.onReceiveStatus (/Users/johngcook/Developer/dex/dex-core/cf_node/functions/node_modules/grpc/src/client_interceptors.js:1209:28)
    at InterceptingListener._callNext (/Users/johngcook/Developer/dex/dex-core/cf_node/functions/node_modules/grpc/src/client_interceptors.js:568:42)
    at InterceptingListener.onReceiveStatus (/Users/johngcook/Developer/dex/dex-core/cf_node/functions/node_modules/grpc/src/client_interceptors.js:618:8)
    at callback (/Users/johngcook/Developer/dex/dex-core/cf_node/functions/node_modules/grpc/src/client_interceptors.js:847:24)

```

My understanding was that using the `initializeAdminApp` feature of `@firebase/testing` eliminated the need for authenticated requests. I am not using Firestore rules of any kind.

#### Steps to reproduce:

I'm running the Firestore emulator in the background with the following command:
```
firebase emulators:start --only firestore
```
with `firebase-cli v7.12.1`.

My test config is like so:

```typescript
// sampleTest.spec.ts
import * as testing from '@firebase/testing';

process.env.FIRESTORE_EMULATOR_HOST = 'localhost:5002';

const projectId = 'testing';
const databaseName = 'firestore-testing';
const databaseURL = 'http://localhost:5002';

const testApp = testing.initializeAdminApp({ projectId, databaseName });
const testFirestore = testApp.firestore();
testFirestore.settings({ experimentalForceLongPolling: true }); // enabling or disabling this line has no effect on the problem

describe('tests', () => {
  beforeAll(() => testing.clearFirestoreData({ projectId })));
  it('should be able to use the firestore emulator', async () => {
    const exampleDoc = await testFirestore.doc('users/exampleUserId').get();
    expect(exampleDoc.exists).toBeFalsy();
  });
});
```

Running the test using `jest@^25.1.0` produces the same error repeatedly.

**Comment by rommelpe** [comment]
Thanks for the report, @the-jackalope. Getting same behavior when using `initializeAdminApp`.

![image](https://user-images.githubusercontent.com/56452310/74334109-3ffa5f80-4dd4-11ea-81f5-83b519f41b5e.png)


Yes, based from [public docs](https://firebase.google.com/docs/firestore/security/test-rules-emulator#run_local_tests), it should be able to get admin-authorized privileges (read & write are allowed). 

> This method returns an initialized admin Firebase app. This app bypasses security rules when performing reads and writes. Use this to create an app authenticated as an admin to set state for tests.

I'm not sure though if the error is really related to Auth permission (which is clearly stated in the error).

I tried using `initializeTestApp` without specified auth variable and was able to read data from the doc.

Changing issue label to get insights from Firestore folks. Thanks

**Comment by wu-hui** [comment]
How did you setup port `5002` as Firestore emulator port? Can you show me your `firebase.json`?

There are couple things that might be contributing to the error:

* You setup both `database` and `firestore` in your tests, they are on the same port. But they are different things: `database` is the Firebase RTDB and `firestore` is Firestore..

* The call to `testFirestore.settings` is still problematic: this overrides the `settings` set by `initializeAdminApp`, which sets up the emulator address and disable `ssl` connection. I think our documentation failed to make this clear.

**Comment by rommelpe** [comment]
Adding my [repro code](https://github.com/rommelpe/quickstart-nodejs/tree/master/firestore-emulator/typescript-quickstart) from quickstart. @the-jackalope feel free to provide yours as well.

**Comment by the-jackalope** [comment]
Hi all, thanks for getting back to me.

@wu-hui :

1. Here is my `firebase.json`:
```json
{
  "functions": {
    "predeploy": [
      "npm --prefix \"$RESOURCE_DIR\" run lint",
      "npm --prefix \"$RESOURCE_DIR\" run build"
    ],
    "source": "functions"
  },
  "emulators": {
    "firestore": {
      "port": 5002
    }
  }
}
```

2. I'm positive that I'm not making a call to `database` anywhere in the codebase. We only use Firestore to hold data and don't touch RTDB in either production or tests. Perhaps not relevant, but we also only start the Firestore emulator from the CLI using the `--only` flag.

3. If I don't call `testFirestore.settings` I get a slightly different error, but am still unable to connect to the emulator:
```
node_modules/@firebase/testing/node_modules/@firebase/logger/dist/index.cjs.js:94
    [2020-02-12T19:56:10.457Z]  @firebase/firestore: Firestore (7.8.1): Could not reach Cloud Firestore backend. Connection failed 1 times. Most recent error: FirebaseError: [code=unavailable]: 14 UNAVAILABLE: failed to connect to all addresses
    This typically indicates that your device does not have a healthy Internet connection at the moment. The client will operate in offline mode until it is able to successfully connect to the backend.

Error: 14 UNAVAILABLE: failed to connect to all addresses
    at Object.exports.createStatusError (/Users/johngcook/Developer/dex/dex-core/cf_node/functions/node_modules/grpc/src/common.js:91:15)
    at Object.onReceiveStatus (/Users/johngcook/Developer/dex/dex-core/cf_node/functions/node_modules/grpc/src/client_interceptors.js:1209:28)
    at InterceptingListener._callNext (/Users/johngcook/Developer/dex/dex-core/cf_node/functions/node_modules/grpc/src/client_interceptors.js:568:42)
    at InterceptingListener.onReceiveStatus (/Users/johngcook/Developer/dex/dex-core/cf_node/functions/node_modules/grpc/src/client_interceptors.js:618:8)
    at callback (/Users/johngcook/Developer/dex/dex-core/cf_node/functions/node_modules/grpc/src/client_interceptors.js:847:24)

```

**Comment by wu-hui** [comment]
@rommelpe Thanks for your repro.

Comment out `testFirestore.settings({ experimentalForceLongPolling: true })` worked for me.

**Comment by wu-hui** [comment]
@the-jackalope 

In the snippet you provided, there are `databaseName` and `databaseURL`. Those are RTDB configuration, not Firestore's.

Can you try removing those? Sorry I was not being very clear earlier.

**Comment by the-jackalope** [comment]
@wu-hui 

Removing the `databaseName`, `databaseURL` has changed something: I no longer receive the `FirebaseError` that is logged from the `@firebase/testing` package. Unfortunately, I still get this error showing up as soon as I try to read/write from the emulator in the test suite:
```
Error: 14 UNAVAILABLE: failed to connect to all addresses
    at Object.exports.createStatusError (/Users/johngcook/Developer/dex/dex-core/cf_node/functions/node_modules/grpc/src/common.js:91:15)
    at Object.onReceiveStatus (/Users/johngcook/Developer/dex/dex-core/cf_node/functions/node_modules/grpc/src/client_interceptors.js:1209:28)
    at InterceptingListener._callNext (/Users/johngcook/Developer/dex/dex-core/cf_node/functions/node_modules/grpc/src/client_interceptors.js:568:42)
    at InterceptingListener.onReceiveStatus (/Users/johngcook/Developer/dex/dex-core/cf_node/functions/node_modules/grpc/src/client_interceptors.js:618:8)
    at callback (/Users/johngcook/Developer/dex/dex-core/cf_node/functions/node_modules/grpc/src/client_interceptors.js:847:24)
```

Appreciate the help!

**Comment by yuchenshi** [comment]
Hi all, I'm late into the conversation but `@firebase/testing` reads `FIRESTORE_EMULATOR_HOST` env var at *import* time, so setting `process.env.FIRESTORE_EMULATOR_HOST` after import won't do anything here. I believe in your case `@firebase/testing` just tries to connect to the default port and fail, resulting the error we're seeing.

For a simple fix, just run the test like `FIRESTORE_EMULATOR_HOST=localhost:5002 jest` (or similar). However, this gives you a _client_ SDK app with admin privileges (with client APIs) rather than an _admin_ SDK app, since the `@firebase/testing` package is intended to help testing rules. I believe in your use case, you probably want to just use `firebase-admin` in your tests, but run it like `firebase emulators:exec --only firestore 'jest'` so your test script connects to the emulator. (Under the hood, `emulators:exec` sets the `FIRESTORE_EMULATOR_HOST` env var for you, which in turn gets automatically picked up by `firebase-admin`).

**Comment by the-jackalope** [comment]
Hi @yuchenshi :

You're right! Setting that env var fixed the connection issues. I also dumped my `process.env` when running my test suite with `firebase emulators:exec` and noticed that in addition to `FIRESTORE_EMULATOR_HOST` being populated with the emulator address, there was another variable `FIREBASE_FIRESTORE_EMULATOR_ADDRESS` which was also being populated with the emulator address.

I'm not sure which Firebase packages rely on the second variable, but I set them both in my Jest config and have had no further connection issues. I appreciate the help and will close this issue.
